mackerel-plugin-mcrouter/lib/mcrouter.go

```diff
@@ -0,0 +1,195 @@
+type McrouterPlugin struct {
+	Prefix    string
+	StatsFile string
+}
+
+// MetricKeyPrefix interface for mackerelplugin
+func (p McrouterPlugin) MetricKeyPrefix() string {
+	if p.Prefix == "" {
+		p.Prefix = "mcrouter"
+	}
+	return p.Prefix
+}
+
+// GraphDefinition interface for mackerelplugin
+func (p McrouterPlugin) GraphDefinition() map[string]mp.Graphs {
+	var cmdMetrics []mp.Metrics
+	for _, name := range cmdMetricNames {
+		cmdMetrics = append(cmdMetrics, mp.Metrics{
+			Name:  name,
+			Label: name,
+			Diff:  true,
+		})
+	}
+
+	var resultMetrics []mp.Metrics
+	for _, name := range resultMetricNames {
+		resultMetrics = append(resultMetrics, mp.Metrics{
+			Name:  name,
+			Label: name,
+			Diff:  true,
+		})
+	}
+
+	labelPrefix := strings.Title(p.Prefix)
+	return map[string]mp.Graphs{
+		"cmd_count": {
+			Label:   (labelPrefix + " Receive Command"),
+			Unit:    "integer",
+			Metrics: cmdMetrics,
+		},
+		"result_count": {
+			Label:   (labelPrefix + " Reply Result"),
+			Unit:    "integer",
+			Metrics: resultMetrics,
+		},
+		"request_processing_time": {
+			Label: (labelPrefix + " Request Processing Time"),
+			Unit:  "float",
+			Metrics: []mp.Metrics{
+				{Name: "duration_us", Label: "Duration(us)", Diff: false},
+			},
+		},
+	}
+}
+
+// FetchMetrics interface for mackerelplugin
+func (p McrouterPlugin) FetchMetrics() (map[string]interface{}, error) {
+	stats, err := readStatsFile(p.StatsFile)
+	if err != nil {
+		return nil, err
+	}
+
+	ret := make(map[string]interface{})
+	statsPrefix := strings.TrimSuffix(path.Base(p.StatsFile), ".stats")
+
+	// Get cmd_[operation]_count stats
+	for _, name := range cmdMetricNames {
+		value := getStats(stats, statsPrefix, name)
+		ret[name], err = strconv.ParseUint(string(value), 10, 64)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Get result_[reply result]_count stats
+	for _, name := range resultMetricNames {
+		value := getStats(stats, statsPrefix, name)
+		ret[name], err = strconv.ParseUint(string(value), 10, 64)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Get duration_us stats
+	for _, name := range []string{"duration_us"} {
+		value := getStats(stats, statsPrefix, name)
+		ret[name], err = strconv.ParseFloat(string(value), 64)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return ret, nil
+}
+
+func readStatsFile(statsFile string) (interface{}, error) {
+	data, err := ioutil.ReadFile(statsFile)
+	if err != nil {
+		return nil, err
+	}
+
+	var stats interface{}
+	d := json.NewDecoder(bytes.NewBuffer(data))
+	d.UseNumber()
+	if err := d.Decode(&stats); err != nil {
+		return nil, err
+	}
+
+	return stats, nil
+}
+
+func getStats(stats interface{}, statsPrefix string, metricName string) json.Number {
+	statsKey := fmt.Sprintf("%s.%s", statsPrefix, metricName)
+	return stats.(map[string]interface{})[statsKey].(json.Number)
+}
+
+// Do the plugin
+func Do() {
+	var (
+		optStatsFile = flag.String("stats-file", "", "Mcrouter stats file")
```

Comment:
It's useful if default (or common) path of the stats file path is set by default.

Reply:
The directory of the stats file has a default value, but there is no default value of the port number included in the file name. So it was not possible to set the default value of `-stats-file` :cry: